tag_tracer.go

```diff
@@ -0,0 +1,300 @@
+		err := t.bumpDeliveryTag(p, topic)
+		if err != nil {
+			log.Warnf("error bumping delivery tag: %s", err)
+		}
+	}
+}
+
+// nearFirstPeers returns the peers who delivered the message while it was still validating
+func (t *tagTracer) nearFirstPeers(msg *Message) []peer.ID {
+	t.Lock()
+	defer t.Unlock()
+	drec := t.deliveries.getRecord(t.msgID(msg.Message))
+	nearFirstPeers := make([]peer.ID, 0, len(drec.peers))
+	// defensive check that this is the first delivery trace -- delivery status should be unknown
+	if drec.status != deliveryUnknown {
+		log.Warnf("unexpected delivery trace: message from %s was first seen %s ago and has delivery status %d", msg.ReceivedFrom, time.Now().Sub(drec.firstSeen), drec.status)
+		return nearFirstPeers
+	}
+
+	drec.status = deliveryValid
+	drec.validated = time.Now()
+
+	for p := range drec.peers {
+		// this check is to make sure a peer can't send us a message twice and get a double count
+		// if it is a first delivery.
+		if p != msg.ReceivedFrom {
+			nearFirstPeers = append(nearFirstPeers, p)
+		}
+	}
+	// we're done with the peers map and can reclaim the memory
+	drec.peers = nil
+	return nearFirstPeers
+}
+
+// -- internalTracer interface methods
+var _ internalTracer = (*tagTracer)(nil)
+
+func (t *tagTracer) AddPeer(p peer.ID, proto protocol.ID) {
+	t.tagPeerIfDirect(p)
+}
+
+func (t *tagTracer) Join(topic string) {
+	t.addDeliveryTag(topic)
+}
+
+func (t *tagTracer) DeliverMessage(msg *Message) {
+	nearFirst := t.nearFirstPeers(msg)
+
+	t.bumpTagsForMessage(msg.ReceivedFrom, msg)
+	for _, p := range nearFirst {
+		t.bumpTagsForMessage(p, msg)
+	}
+}
+
+func (t *tagTracer) Leave(topic string) {
+	t.removeDeliveryTag(topic)
+}
+
+func (t *tagTracer) Graft(p peer.ID, topic string) {
+	t.tagMeshPeer(p, topic)
+}
+
+func (t *tagTracer) Prune(p peer.ID, topic string) {
+	t.untagMeshPeer(p, topic)
+}
+
+func (t *tagTracer) ValidateMessage(msg *Message) {
+	t.Lock()
+	defer t.Unlock()
+
+	// create a delivery record for the message
+	_ = t.deliveries.getRecord(t.msgID(msg.Message))
+}
+
+func (t *tagTracer) DuplicateMessage(msg *Message) {
+	t.Lock()
+	defer t.Unlock()
+
+	drec := t.deliveries.getRecord(t.msgID(msg.Message))
+	if drec.status == deliveryUnknown {
+		// the message is being validated; track the peer delivery and wait for
+		// the Deliver/Reject notification.
+		drec.peers[msg.ReceivedFrom] = struct{}{}
+	}
+}
+
+func (t *tagTracer) RejectMessage(msg *Message, reason string) {
+	t.Lock()
+	defer t.Unlock()
+
+	// mark message as invalid and release tracking info
+	drec := t.deliveries.getRecord(t.msgID(msg.Message))
+
+	// defensive check that this is the first rejection trace -- delivery status should be unknown
+	if drec.status != deliveryUnknown {
+		log.Warnf("unexpected rejection trace: message from %s was first seen %s ago and has delivery status %d", msg.ReceivedFrom, time.Now().Sub(drec.firstSeen), drec.status)
+		return
+	}
+
+	drec.status = deliveryInvalid
+	drec.peers = nil
+}
```

Comment:
actually, there are some cases where the action (marking as invalid) is incorrect; if the rejection is because of invalid or missing signature or validation queue full we shouldn't clear the delivery record. This is because the rejection happens without entering the validation pipeline and we can end up rejecting a message that is still validating.

So let's add a check here for the reason and short-circuit if it is one of the above; see also the handling in score.go

Reply:
Might even make sense to check explicitly for one of validation throttled, validation ignored, validation rejected as these are the only ones that indicate a message that was in the validation queue.
Then we can set the delivery status to a known value (not necessarily invalid).